cmd/soroban-rpc/internal/ledgerentry_storage/db.go

```diff
@@ -0,0 +1,295 @@
+package ledgerentry_storage
+
+import (
+	"context"
+	"database/sql"
+	"embed"
+	"fmt"
+	"strconv"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/jmoiron/sqlx"
+	_ "github.com/mattn/go-sqlite3"
+	migrate "github.com/rubenv/sql-migrate"
+	"github.com/stellar/go/support/errors"
+	"github.com/stellar/go/xdr"
+)
+
+//go:embed migrations/*.sql
+var migrations embed.FS
+
+var ErrEmptyDB = errors.New("DB is empty")
+
+const (
+	ledgerEntriesTableName      = "ledger_entries"
+	ledgerEntriesMetaTableName  = "ledger_entries_meta"
+	latestLedgerSequenceMetaKey = "LatestLedgerSequence"
+)
+
+type DB interface {
+	LedgerEntryStorage
+	GetLatestLedgerSequence() (uint32, error)
+	NewLedgerEntryUpdaterTx(forLedgerSequence uint32, maxBatchSize int) (LedgerEntryUpdaterTx, error)
+}
+
+type LedgerEntryUpdaterTx interface {
+	UpsertLedgerEntry(key xdr.LedgerKey, entry xdr.LedgerEntry) error
+	DeleteLedgerEntry(key xdr.LedgerKey) error
+	Done() error
+}
+
+type sqlDB struct {
+	db *sqlx.DB
+}
+
+func OpenSQLiteDB(dbFilePath string) (DB, error) {
+	db, err := sqlx.Open("sqlite3", dbFilePath)
+	if err != nil {
+		return nil, errors.Wrap(err, "open failed")
+	}
+
+	ret := &sqlDB{
+		db: db,
+	}
+
+	err = runMigrations(ret.db.DB, "sqlite3")
+
+	return ret, err
+}
+
+func getLedgerEntry(tx *sqlx.Tx, buffer *xdr.EncodingBuffer, key xdr.LedgerKey) (xdr.LedgerEntry, error) {
+	encodedKey, err := encodeLedgerKey(buffer, key)
+	if err != nil {
+		return xdr.LedgerEntry{}, err
+	}
+
+	sqlStr, args, err := sq.Select("entry").From(ledgerEntriesTableName).Where(sq.Eq{"key": encodedKey}).ToSql()
+	if err != nil {
+		return xdr.LedgerEntry{}, err
+	}
+	var results []string
+	if err := tx.Select(&results, sqlStr, args...); err != nil {
+		return xdr.LedgerEntry{}, err
+	}
+	if len(results) != 1 {
+		return xdr.LedgerEntry{}, sql.ErrNoRows
+	}
+	ledgerEntryBase64 := results[0]
+	var result xdr.LedgerEntry
+	err = xdr.SafeUnmarshalBase64(ledgerEntryBase64, &result)
+	return result, err
+}
+
+func flushLedgerEntryBatch(tx *sqlx.Tx, encodedKeyEntries map[string]*string) error {
+	upsertCount := 0
+	upsertSQL := sq.Replace(ledgerEntriesTableName)
+	var deleteKeys = make([]interface{}, 0, len(encodedKeyEntries))
+	for key, entry := range encodedKeyEntries {
+		if entry != nil {
+			upsertSQL = upsertSQL.Values(key, entry)
+			upsertCount += 1
+		} else {
+			deleteKeys = append(deleteKeys, interface{}(key))
+		}
+	}
+
+	if upsertCount > 0 {
+		sqlStr, args, err := upsertSQL.ToSql()
+		if err != nil {
+			return err
+		}
+		if _, err = tx.Exec(sqlStr, args...); err != nil {
+			return err
+		}
+	}
+
+	if len(deleteKeys) > 0 {
+		sqlStr, args, err := sq.Delete(ledgerEntriesTableName).Where(sq.Eq{"key": deleteKeys}).ToSql()
+		if err != nil {
+			return err
+		}
+		_, err = tx.Exec(sqlStr, args...)
```

Comment:
duplicated line ?

Reply:
Deletions and updates are queried separately